src/mlpack/tests/recurrent_network_test.cpp

```diff
@@ -868,3 +868,67 @@ TEST_CASE("LargeRhoValueRnnTest", "[RecurrentNetworkTest]")
   model.Train(inputs[0], targets[0], opt);
   INFO("Training over");
 }
+
+/**
+ * Test to make sure that an error is thrown when input with
+ * wrong input shape is provided to a RNN.
+ */
+TEST_CASE("RNNCheckInputShapeTest", "[RecurrentNetworkTest]")
+{
+  const size_t rho = 10;
+
+  // Generate 12 (2 * 6) noisy sines. A single sine contains rho
+  // points/features.
+  arma::cube input;
+  arma::mat labelsTemp;
+  GenerateNoisySines(input, labelsTemp, rho, 6);
+
+  arma::cube labels = arma::zeros<arma::cube>(1, labelsTemp.n_cols, rho);
+  for (size_t i = 0; i < labelsTemp.n_cols; ++i)
+  {
+    const int value = arma::as_scalar(arma::find(
+        arma::max(labelsTemp.col(i)) == labelsTemp.col(i), 1)) + 1;
+    labels.tube(0, i).fill(value);
+  }
+
+  /**
+   * Construct a network with 1 input unit, 4 hidden units and 10 output
+   * units. The hidden layer is connected to itself. The network structure
+   * looks like:
+   *
+   *  Input         Hidden        Output
+   * Layer(1)      Layer(4)      Layer(10)
+   * +-----+       +-----+       +-----+
+   * |     |       |     |       |     |
+   * |     +------>|     +------>|     |
+   * |     |    ..>|     |       |     |
+   * +-----+    .  +--+--+       +-----+
+   *            .     .
+   *            .     .
+   *            .......
+   */
+  Add<> add(4);
+  // Purposely providing wrong input shape of 3.
+  // The correct input shape is 1.
+  Linear<> lookup(3, 4);
+  SigmoidLayer<> sigmoidLayer;
+  Linear<> linear(4, 4);
+  Recurrent<>* recurrent = new Recurrent<>(add, lookup, linear,
+      sigmoidLayer, rho);
+
+  RNN<> model(rho);
+  model.Add<IdentityLayer<> >();
+  model.Add(recurrent);
+  model.Add<Linear<> >(4, 10);
+  model.Add<LogSoftMax<> >();
+
+  std::string expectedMsg = "RNN<>::Train(): ";
+              expectedMsg += "the first layer of the network expects ";
+              expectedMsg += std::to_string(3) + " elements, ";
+              expectedMsg += "but the input has " + std::to_string(1) + " dimensions! ";
```

Comment:
Personally I think it would be fine just to match the exception type here; we might change the internal output format over time, and it would be tedious to have to update these tests every time. :+1:

(This applies to the FFN test too.)

Reply:
I did this because I thought it will be better to add an extra layer of verification that also tests which message is thrown but I see how tedious updating this would be every time message is changed. I will change it 👍